<a href="https://colab.research.google.com/github/sylkip3/Automated_Recon/blob/main/Portugal_Association_Rules_v2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# T5.1 — Association Rule Mining: Portugal Online Retail

We apply the Apriori algorithm to Portugal's online retail data.

Reference documentation:
- https://rasbt.github.io/mlxtend/user_guide/frequent_patterns/apriori/
- https://rasbt.github.io/mlxtend/user_guide/frequent_patterns/association_rules/

##DATASET DESCRIPTION
The Portugal dataset represents online purchases that were made by customers based in Portugal. The dataset size has 58 transactions for 713 products indicating that most products appear in fewer transactions. A look at the dataset indicates that the dataset in already pre-processed.

In [16]:
# Source - https://stackoverflow.com/a/66891328
# Posted by caesarsol
# Retrieved 2026-02-25, License - CC BY-SA 4.0

# Avoids scroll-in-the-scroll in the entire Notebook
from IPython.display import Javascript
def resize_colab_cell():
  display(Javascript('google.colab.output.setIframeHeight(0, true, {maxHeight: 5000})'))
get_ipython().events.register('pre_run_cell', resize_colab_cell)



In [17]:
import warnings
warnings.filterwarnings('ignore')
import warnings
import numpy as np
import pandas as pd
from mlxtend.frequent_patterns import apriori, association_rules

<IPython.core.display.Javascript object>

In [18]:
# Load the Portugal basket dataset and remove the transaction ID column

basket_Portugal = pd.read_csv('Portugal_online_retail.csv')
basket_Portugal = basket_Portugal.drop('InvoiceNo', axis=1)
basket_Portugal

<IPython.core.display.Javascript object>

,10 COLOUR SPACEBOY PEN,12 PENCIL SMALL TUBE WOODLAND,12 PENCILS SMALL TUBE RED RETROSPOT,12 PENCILS TALL TUBE POSY,12 PENCILS TALL TUBE RED RETROSPOT,12 PENCILS TALL TUBE WOODLAND,20 DOLLY PEGS RETROSPOT,3 PIECE SPACEBOY COOKIE CUTTER SET,3 STRIPEY MICE FELTCRAFT,36 FOIL HEART CAKE CASES,...,WRAP FLOWER SHOP,WRAP GINGHAM ROSE,WRAP PAISLEY PARK,WRAP PINK FAIRY CAKES,WRAP RED APPLES,WRAP RED VINTAGE DOILY,WRAP SUKI AND FRIENDS,YOU'RE CONFUSING ME METAL SIGN,ZINC FOLKART SLEIGH BELLS,ZINC WIRE KITCHEN ORGANISER
0,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
1,0,0,0,0,0,0,0,0,1,0,...,0,0,0,0,0,0,0,0,0,0
2,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
3,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
4,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
5,0,0,0,0,1,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
6,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
7,1,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
8,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
9,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0


In [19]:
# Find frequent itemsets — only keep itemsets appearing in >= 10% of transactions
frq_items = apriori(basket_Portugal, min_support=0.1, use_colnames=True)
frq_items.sort_values(by=['support'], ascending=False)

<IPython.core.display.Javascript object>

,support,itemsets
25,0.517241,(POSTAGE)
14,0.241379,(LUNCH BAG RED RETROSPOT)
30,0.241379,(RETROSPOT TEA SET CERAMIC 11 PC)
1,0.206897,(BAKING SET 9 PIECE RETROSPOT)
11,0.206897,(LUNCH BAG CARS BLUE)
...,...,...
103,0.103448,"(JUMBO BAG SCANDINAVIAN BLUE PAISLEY, JUMBO BA..."
105,0.103448,"(JUMBO BAG RED RETROSPOT, JUMBO BAG PINK VINTA..."
106,0.103448,"(JUMBO BAG SCANDINAVIAN BLUE PAISLEY, JUMBO BA..."
107,0.103448,"(JUMBO BAG SCANDINAVIAN BLUE PAISLEY, JUMBO BA..."


In [20]:
# --- Confidence Level 1: Low (0.3) ---
# Lower threshold — returns more rules, including weaker associations
rules_30 = association_rules(frq_items, metric='confidence', min_threshold=0.3)
print(f'min_confidence=0.3 → {len(rules_30)} rules')
rules_30.sort_values(by=['lift'], ascending=False)

<IPython.core.display.Javascript object>

min_confidence=0.3 → 317 rules


,antecedents,consequents,antecedent support,consequent support,support,confidence,lift,representativity,leverage,conviction,zhangs_metric,jaccard,certainty,kulczynski
284,"(LUNCH BAG PINK POLKADOT, JUMBO SHOPPER VINTAG...","(JUMBO BAG SCANDINAVIAN BLUE PAISLEY, JUMBO BA...",0.103448,0.137931,0.103448,1.000000,7.250000,1.0,0.089180,inf,0.961538,0.750000,1.000000,0.875000
281,"(JUMBO BAG SCANDINAVIAN BLUE PAISLEY, JUMBO SH...","(LUNCH BAG PINK POLKADOT, JUMBO BAG PINK VINTA...",0.137931,0.103448,0.103448,0.750000,7.250000,1.0,0.089180,3.586207,1.000000,0.750000,0.721154,0.875000
279,"(JUMBO BAG SCANDINAVIAN BLUE PAISLEY, JUMBO BA...","(LUNCH BAG PINK POLKADOT, JUMBO SHOPPER VINTAG...",0.137931,0.103448,0.103448,0.750000,7.250000,1.0,0.089180,3.586207,1.000000,0.750000,0.721154,0.875000
282,"(LUNCH BAG PINK POLKADOT, JUMBO BAG PINK VINTA...","(JUMBO BAG SCANDINAVIAN BLUE PAISLEY, JUMBO SH...",0.103448,0.137931,0.103448,1.000000,7.250000,1.0,0.089180,inf,0.961538,0.750000,1.000000,0.875000
66,(SCANDINAVIAN PAISLEY PICNIC BAG),(PINK VINTAGE PAISLEY PICNIC BAG),0.120690,0.137931,0.120690,1.000000,7.250000,1.0,0.104043,inf,0.980392,0.875000,1.000000,0.937500
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
61,(LUNCH BAG RED RETROSPOT),(RETROSPOT TEA SET CERAMIC 11 PC),0.241379,0.241379,0.103448,0.428571,1.775510,1.0,0.045184,1.327586,0.575758,0.272727,0.246753,0.428571
62,(RETROSPOT TEA SET CERAMIC 11 PC),(LUNCH BAG RED RETROSPOT),0.241379,0.241379,0.103448,0.428571,1.775510,1.0,0.045184,1.327586,0.575758,0.272727,0.246753,0.428571
79,(RECIPE BOX RETROSPOT),(POSTAGE),0.137931,0.517241,0.103448,0.750000,1.450000,1.0,0.032105,1.931034,0.360000,0.187500,0.482143,0.475000
0,(BAKING SET 9 PIECE RETROSPOT),(POSTAGE),0.206897,0.517241,0.137931,0.666667,1.288889,1.0,0.030916,1.448276,0.282609,0.235294,0.309524,0.466667


In [21]:
# --- Confidence Level 2: Medium (0.6) ---
# Balanced threshold — filters out weaker rules while keeping moderate associations
rules_60 = association_rules(frq_items, metric='confidence', min_threshold=0.6)
print(f'min_confidence=0.6 → {len(rules_60)} rules')
rules_60.sort_values(by=['lift'], ascending=False)

<IPython.core.display.Javascript object>

min_confidence=0.6 → 258 rules


,antecedents,consequents,antecedent support,consequent support,support,confidence,lift,representativity,leverage,conviction,zhangs_metric,jaccard,certainty,kulczynski
230,"(LUNCH BAG PINK POLKADOT, JUMBO SHOPPER VINTAG...","(JUMBO BAG SCANDINAVIAN BLUE PAISLEY, JUMBO BA...",0.103448,0.137931,0.103448,1.000000,7.250000,1.0,0.089180,inf,0.961538,0.750000,1.000000,0.875000
45,(SCANDINAVIAN PAISLEY PICNIC BAG),(PINK VINTAGE PAISLEY PICNIC BAG),0.120690,0.137931,0.120690,1.000000,7.250000,1.0,0.104043,inf,0.980392,0.875000,1.000000,0.937500
225,"(JUMBO BAG SCANDINAVIAN BLUE PAISLEY, JUMBO BA...","(LUNCH BAG PINK POLKADOT, JUMBO SHOPPER VINTAG...",0.137931,0.103448,0.103448,0.750000,7.250000,1.0,0.089180,3.586207,1.000000,0.750000,0.721154,0.875000
227,"(JUMBO BAG SCANDINAVIAN BLUE PAISLEY, JUMBO SH...","(LUNCH BAG PINK POLKADOT, JUMBO BAG PINK VINTA...",0.137931,0.103448,0.103448,0.750000,7.250000,1.0,0.089180,3.586207,1.000000,0.750000,0.721154,0.875000
228,"(LUNCH BAG PINK POLKADOT, JUMBO BAG PINK VINTA...","(JUMBO BAG SCANDINAVIAN BLUE PAISLEY, JUMBO SH...",0.103448,0.137931,0.103448,1.000000,7.250000,1.0,0.089180,inf,0.961538,0.750000,1.000000,0.875000
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
5,(CHARLOTTE BAG SUKI DESIGN),(RETROSPOT TEA SET CERAMIC 11 PC),0.172414,0.241379,0.103448,0.600000,2.485714,1.0,0.061831,1.896552,0.722222,0.333333,0.472727,0.514286
42,(RED RETROSPOT CHARLOTTE BAG),(LUNCH BAG RED RETROSPOT),0.172414,0.241379,0.103448,0.600000,2.485714,1.0,0.061831,1.896552,0.722222,0.333333,0.472727,0.514286
56,(PLASTERS IN TIN VINTAGE PAISLEY),(RETROSPOT TEA SET CERAMIC 11 PC),0.172414,0.241379,0.103448,0.600000,2.485714,1.0,0.061831,1.896552,0.722222,0.333333,0.472727,0.514286
57,(RECIPE BOX RETROSPOT),(POSTAGE),0.137931,0.517241,0.103448,0.750000,1.450000,1.0,0.032105,1.931034,0.360000,0.187500,0.482143,0.475000


In [22]:
# --- Confidence Level 3: High (0.9) --- SELECTED
# Strictest threshold — only retains rules where the antecedent implies
# the consequent in at least 90% of transactions, removing noise


rules_90 = association_rules(frq_items, metric='confidence', min_threshold=0.9)
print(f'min_confidence=0.9 → {len(rules_90)} rules  ← SELECTED')
rules_90.sort_values(by=['lift'], ascending=False)

<IPython.core.display.Javascript object>

min_confidence=0.9 → 47 rules  ← SELECTED


,antecedents,consequents,antecedent support,consequent support,support,confidence,lift,representativity,leverage,conviction,zhangs_metric,jaccard,certainty,kulczynski
1,(SCANDINAVIAN PAISLEY PICNIC BAG),(PINK VINTAGE PAISLEY PICNIC BAG),0.120690,0.137931,0.120690,1.0,7.250000,1.0,0.104043,inf,0.980392,0.875000,1.0,0.937500
23,"(PLASTERS IN TIN VINTAGE PAISLEY, PLASTERS IN ...",(PLASTERS IN TIN WOODLAND ANIMALS),0.103448,0.137931,0.103448,1.0,7.250000,1.0,0.089180,inf,0.961538,0.750000,1.0,0.875000
22,"(LUNCH BAG RED RETROSPOT, SCANDINAVIAN PAISLEY...",(PINK VINTAGE PAISLEY PICNIC BAG),0.103448,0.137931,0.103448,1.0,7.250000,1.0,0.089180,inf,0.961538,0.750000,1.0,0.875000
41,"(LUNCH BAG PINK POLKADOT, JUMBO SHOPPER VINTAG...","(JUMBO BAG SCANDINAVIAN BLUE PAISLEY, JUMBO BA...",0.103448,0.137931,0.103448,1.0,7.250000,1.0,0.089180,inf,0.961538,0.750000,1.0,0.875000
40,"(LUNCH BAG PINK POLKADOT, JUMBO BAG PINK VINTA...","(JUMBO BAG SCANDINAVIAN BLUE PAISLEY, JUMBO SH...",0.103448,0.137931,0.103448,1.0,7.250000,1.0,0.089180,inf,0.961538,0.750000,1.0,0.875000
5,"(JUMBO BAG RED RETROSPOT, JUMBO SHOPPER VINTAG...",(JUMBO BAG PINK VINTAGE PAISLEY),0.120690,0.155172,0.120690,1.0,6.444444,1.0,0.101962,inf,0.960784,0.777778,1.0,0.888889
3,"(JUMBO BAG SCANDINAVIAN BLUE PAISLEY, JUMBO BA...",(JUMBO BAG PINK VINTAGE PAISLEY),0.103448,0.155172,0.103448,1.0,6.444444,1.0,0.087396,inf,0.942308,0.666667,1.0,0.833333
11,"(JUMBO BAG SCANDINAVIAN BLUE PAISLEY, LUNCH BA...",(JUMBO BAG PINK VINTAGE PAISLEY),0.120690,0.155172,0.120690,1.0,6.444444,1.0,0.101962,inf,0.960784,0.777778,1.0,0.888889
14,"(LUNCH BAG PINK POLKADOT, JUMBO SHOPPER VINTAG...",(JUMBO BAG PINK VINTAGE PAISLEY),0.103448,0.155172,0.103448,1.0,6.444444,1.0,0.087396,inf,0.942308,0.666667,1.0,0.833333
28,"(JUMBO BAG SCANDINAVIAN BLUE PAISLEY, JUMBO BA...","(JUMBO BAG PINK VINTAGE PAISLEY, JUMBO SHOPPER...",0.103448,0.155172,0.103448,1.0,6.444444,1.0,0.087396,inf,0.942308,0.666667,1.0,0.833333


In [23]:
# Top 3 most important rules for Portugal, ranked by lift
# Lift > 1 means the items are more likely to be bought together than by chance
top3 = rules_90.sort_values(by=['lift'], ascending=False).head(3).reset_index(drop=True)
top3

<IPython.core.display.Javascript object>

,antecedents,consequents,antecedent support,consequent support,support,confidence,lift,representativity,leverage,conviction,zhangs_metric,jaccard,certainty,kulczynski
0,(SCANDINAVIAN PAISLEY PICNIC BAG),(PINK VINTAGE PAISLEY PICNIC BAG),0.120690,0.137931,0.120690,1.0,7.25,1.0,0.104043,inf,0.980392,0.875,1.0,0.9375
1,"(PLASTERS IN TIN VINTAGE PAISLEY, PLASTERS IN ...",(PLASTERS IN TIN WOODLAND ANIMALS),0.103448,0.137931,0.103448,1.0,7.25,1.0,0.089180,inf,0.961538,0.750,1.0,0.8750
2,"(LUNCH BAG RED RETROSPOT, SCANDINAVIAN PAISLEY...",(PINK VINTAGE PAISLEY PICNIC BAG),0.103448,0.137931,0.103448,1.0,7.25,1.0,0.089180,inf,0.961538,0.750,1.0,0.8750


## Report — Portugal

**Confidence level selected: 0.9**

With 58 transactions and min_support set to 0.1, three confidence thresholds were tested on the dataset:
- **0.3** — characterised by a large number of rules. Many of the rules were identified as weak and unlikely to show genuine purchasing behaviour.
- **0.6** — indicates a more balanced set of rules.It however includes associations that are unreliable enough to be acted on commercially.
- **0.9** — This threshold was selected as it produces the most actionable and trustworthy  insights with the buying pattern holding in at least 90% of the cases. Retains only the strongest rules.
---

**Rule 1: {SCANDINAVIAN PAISLEY PICNIC BAG} → {PINK VINTAGE PAISLEY PICNIC BAG}**  
Support = 0.1207, Confidence = 1.0, Lift = 7.25
Customers who bought the Scandinavian Paisley Picnic Bag also bought the Pink Vintage Paisley Picnic Bag with no exception. The lift of 7.25 means this co-purchase occurs 7.25 times more likely than if the two products were independently bought. The two items are effectively  purchased as a set, and the retailer should put them together.

---

**Rule 2: {PLASTERS IN TIN CIRCUS PARADE, PLASTERS IN TIN VINTAGE PAISLEY} → {PLASTERS IN TIN WOODLAND ANIMALS}**  
Support = 0.1034, Confidence = 1.0, Lift = 7.25

Customers who bought, both the Circus Parade and Vintage Paisley plaster tins also bought the Woodland Animals tin. This implies Portuguese customers tend to collect the full Plasters in Tin range instead of buying a single design. For this kind of market,a bundled three-pack offer or a 'complete the collection' prompt would be a great and effective upselling strategy.

---

**Rule 3: {SCANDINAVIAN PAISLEY PICNIC BAG, LUNCH BAG RED RETROSPOT} → {PINK VINTAGE PAISLEY PICNIC BAG}**  
Support = 0.1034, Confidence = 1.0, Lift = 7.25

Portuguese customers who had both the Scandinavian Paisley Picnic Bag and the Red Retrospot Lunch Bag in their basket also added the Pink Vintage Paisley Picnic bag. This extends Rule 1 and reveals a distinct Portuguese customer segment that gravitates toward coordinated paisley-patterned bag sets. Recommending the Pink Vintage Paisley bag whenever any combination of these items are in in the cart would be appropriate.